pkg/instance/instance_integration_test.go

```diff
@@ -61,68 +71,143 @@ func TestInstanceHandler(t *testing.T) {
 	groupService := groupService{group: group}
 	stacks := stack.Stacks{
 		"whoami-go": stack.WhoamiGo,
+		"dhis2":     stack.DHIS2,
 	}
 	stackService := stack.NewService(stacks)
 	// classification 'test' does not actually exist, this is used to decrypt the stack parameters
 	helmfileService := instance.NewHelmfileService("../../stacks", stackService, "test")
 	instanceService := instance.NewService(instanceRepo, groupService, stackService, helmfileService)
 
+	s3Dir := t.TempDir()
+	s3Bucket := "database-bucket"
+	err = os.Mkdir(s3Dir+"/"+s3Bucket, 0o755)
+	require.NoError(t, err, "failed to create S3 output bucket")
+	s3 := inttest.SetupS3(t, s3Dir)
+	uploader := manager.NewUploader(s3.Client)
+	s3Client := storage.NewS3Client(s3.Client, uploader)
+	databaseRepository := database.NewRepository(db)
+	databaseService := database.NewService(s3Bucket, s3Client, groupService, databaseRepository)
+
+	authenticator := func(ctx *gin.Context) {
+		ctx.Set("user", &model.User{
+			ID:    user.ID, // TODO: Why is this necessary all of a sudden?
+			Email: "user1@dhis2.org",
+			Groups: []model.Group{
+				{
+					Name: "group-name",
+				},
+			},
+		})
+	}
 	client := inttest.SetupHTTPServer(t, func(engine *gin.Engine) {
 		var twoDayTTL uint = 172800
 		instanceHandler := instance.NewHandler(groupService, instanceService, twoDayTTL)
-		instance.Routes(engine, func(ctx *gin.Context) {
-			ctx.Set("user", user)
-		}, instanceHandler)
+		instance.Routes(engine, authenticator, instanceHandler)
+
+		databaseHandler := database.NewHandler(databaseService, groupService, instanceService, stackService)
+		database.Routes(engine, authenticator, databaseHandler)
 	})
 
 	t.Run("Deploy", func(t *testing.T) {
 		var instance model.Instance
 		client.PostJSON(t, "/instances", strings.NewReader(`{
 			"name": "test-whoami",
-			"groupName": "test",
+			"groupName": "group-name",
 			"stackName": "whoami-go"
 		}`), &instance, inttest.WithAuthToken("sometoken"))
 
-		ctx, cancel := context.WithCancel(context.Background())
-		watch, err := k8sClient.Client.CoreV1().Pods(group.Name).Watch(ctx, metav1.ListOptions{
-			LabelSelector: "app.kubernetes.io/instance=" + instance.Name,
-		})
-		require.NoErrorf(t, err, "failed to find pod for instance %q", instance.Name)
-
-		timeout := 20 * time.Second
-		tm := time.NewTimer(timeout)
-		defer tm.Stop()
-		for {
-			select {
-			case <-tm.C:
-				assert.Fail(t, "timed out waiting on pod")
+		assertPodIsReady(t, k8sClient, group.Name, instance.Name)
+	})
+
+	var databaseID string
+	{
+		t.Log("Upload")
+		var b bytes.Buffer
+		w := multipart.NewWriter(&b)
+		err := w.WriteField("group", "group-name")
+		require.NoError(t, err, "failed to write form field")
+		f, err := w.CreateFormFile("database", "mydb")
+		require.NoError(t, err, "failed to create form file")
+		// TODO: Why does this work?
```

Comment:
What do you mean? https://pkg.go.dev/mime/multipart#Writer.CreateFormFile and https://pkg.go.dev/mime/multipart#Writer.CreatePart should explain how it works

Reply:
I'm surprised the seed script doesn't return an error when we're seeding a file with the content "file contents"

Reply:
Ah ok, well then you need to investigate 😋